cranelift/isle/isle/src/serialize.rs

```diff
@@ -0,0 +1,523 @@
+            ) {
+                continue;
+            }
+
+            // TODO: proactively put some bindings in `Emitted` state
+            // That makes them visible to the best-binding heuristic, which prefers to match on
+            // already-emitted bindings first. This helps to sort cheap computations before
+            // expensive ones.
+
+            let idx = idx.try_into().unwrap();
+            if self.ready(idx) < Readiness::Available {
+                if binding
+                    .sources()
+                    .iter()
+                    .all(|&source| self.ready(source) >= Readiness::Available)
+                {
+                    self.set_ready(idx, Readiness::Available);
+                }
+            }
+        }
+    }
+
+    fn use_expr(&mut self, name: BindingId) {
+        if self.ready(name) < Readiness::Emitted {
+            self.set_ready(name, Readiness::Emitted);
+            let binding = &self.rules.bindings[name.index()];
+            for &source in binding.sources() {
+                self.use_expr(source);
+            }
+            let let_bind = match binding {
+                // Never let-bind trivial expressions.
+                Binding::MatchTuple { .. } => false,
+                Binding::ConstInt { .. } => false,
+                Binding::ConstPrim { .. } => false,
+                Binding::Argument { .. } => false,
+                // Only let-bind variant constructors if they have some fields. Building a
+                // variant with no fields is cheap, but don't duplicate more complex
+                // expressions.
+                Binding::MakeVariant { fields, .. } => !fields.is_empty(),
+                _ => true,
+            };
+            if let_bind {
+                self.bind_order.push(name);
+            }
+        }
+    }
+
+    fn ready(&self, source: BindingId) -> Readiness {
+        self.ready[source.index()]
+    }
+
+    fn set_ready(&mut self, source: BindingId, state: Readiness) {
+        let old = &mut self.ready[source.index()];
+        debug_assert!(*old <= state);
+
+        // Add candidates for this binding, but only when it first becomes available.
+        if let Readiness::Unavailable = old {
+            self.candidates.extend(
+                // A binding site will have at most one of these kinds of constraint, and many have
+                // none. But `best_single_binding` has to check all candidates anyway, so let it
+                // figure out which (if any) of these are applicable. It will only check false
+                // candidates once on any partition, removing them from this list immediately.
+                [
+                    MatchKind::Constraint(source),
+                    MatchKind::Iterator(source),
+                    // Obviously this binding site equals itself, so this checks whether it
+                    // participates in any equality constraints at all.
+                    MatchKind::Equal(source, source),
```

Comment:
This is super-confusing and it seems maybe a different representation is in order: why should there ever be a test of `source == source` present? The explanation given seems to indicate that this triggers some other check; if so can we make that explicit?